File: projects/sdk-kotlin/src/main/kotlin/orca/sdk/Manifest.kt
// Plugin manifest — `orca-plugin.toml`. Wire-compatible with
// projects/sdk/src/manifest.rs and the other language ports.
package orca.sdk

import java.nio.file.Files
import java.nio.file.Path
import org.tomlj.Toml
import org.tomlj.TomlParseResult
import org.tomlj.TomlTable

const val MANIFEST_FILENAME: String = "orca-plugin.toml"

data class PluginSection(
    val id: String,
    val version: String,
    val minOrcaVersion: String,
)

data class RuntimeSection(
    val binary: String?,
    val image: String?,
    val mode: String,
    val eager: Boolean,
)

data class SurfacesSection(
    val mcp: Boolean = false,
    val cli: Boolean = false,
    val ui: Boolean = false,
    val docs: Boolean = false,
    val jobs: Boolean = false,
    val storage: Boolean = false,
    val federation: Boolean = false,
)

data class CapabilityDecl(val name: String, val sensitivity: String)

data class Manifest(
    val plugin: PluginSection,
    val runtime: RuntimeSection,
    val surfaces: SurfacesSection,
    val capabilities: List<CapabilityDecl>,
)

private val ALLOWED_TOP = setOf("plugin", "runtime", "surfaces", "capabilities")
private val ALLOWED_PLUGIN = setOf("id", "version", "min_orca_version")
private val ALLOWED_RUNTIME = setOf("binary", "image", "mode", "eager")
private val ALLOWED_SURFACES = setOf("mcp", "cli", "ui", "docs", "jobs", "storage", "federation")
private val ALLOWED_CAPABILITY = setOf("name", "sensitivity")

object ManifestParser {
    fun parseString(s: String): Manifest {
        val parsed: TomlParseResult = Toml.parse(s)
        if (parsed.hasErrors()) {
            error("parse $MANIFEST_FILENAME: ${parsed.errors().joinToString("; ")}")
        }
        denyUnknown(parsed.keySet(), ALLOWED_TOP, "top-level")

        val pluginTable = parsed.getTable("plugin") ?: error("missing [plugin]")
        denyUnknown(pluginTable.keySet(), ALLOWED_PLUGIN, "[plugin]")
        val id = (pluginTable.getString("id") ?: "").trim()
        if (id.isEmpty()) error("plugin.id must not be empty")
        for (ch in id) {
            if (ch == ' ' || ch == '\t' || ch == '\n' || ch == '/' || ch == '\\') {
                error("plugin.id \"$id\" contains invalid characters (whitespace or path separators)")
            }
        }
        val version = pluginTable.getString("version") ?: error("plugin.version missing")
        val minOrca = pluginTable.getString("min_orca_version") ?: error("plugin.min_orca_version missing")
        checkSemver(version, "plugin.version")
        checkSemver(minOrca, "plugin.min_orca_version")

        val runtimeTable = parsed.getTable("runtime") ?: error("missing [runtime]")
        denyUnknown(runtimeTable.keySet(), ALLOWED_RUNTIME, "[runtime]")
        val binary = runtimeTable.getString("binary")
        val image = runtimeTable.getString("image")
        if (binary != null && image != null) error("runtime.binary and runtime.image are mutually exclusive")
        if (binary == null && image == null) error("runtime requires either `binary` or `image`")
        val mode = runtimeTable.getString("mode") ?: "process"
        if (mode != "process") error("runtime.mode \"$mode\": only \"process\" is supported in v0")
        val eager = runtimeTable.getBoolean("eager") ?: false

        val surfacesTable = parsed.getTable("surfaces")
        if (surfacesTable != null) denyUnknown(surfacesTable.keySet(), ALLOWED_SURFACES, "[surfaces]")
        val surfaces = SurfacesSection(
            mcp = surfacesTable?.getBoolean("mcp") ?: false,
            cli = surfacesTable?.getBoolean("cli") ?: false,
            ui = surfacesTable?.getBoolean("ui") ?: false,
            docs = surfacesTable?.getBoolean("docs") ?: false,
            jobs = surfacesTable?.getBoolean("jobs") ?: false,
            storage = surfacesTable?.getBoolean("storage") ?: false,
            federation = surfacesTable?.getBoolean("federation") ?: false,
        )

        val capsArray = parsed.getArray("capabilities")
        val seen = HashSet<String>()
        val capabilities = buildList {
            if (capsArray != null) {
                for (i in 0 until capsArray.size()) {
                    val tbl = capsArray.getTable(i)
                    denyUnknown(tbl.keySet(), ALLOWED_CAPABILITY, "[[capabilities]]")
                    val name = (tbl.getString("name") ?: "").trim()
                    if (name.isEmpty()) error("capability.name must not be empty")
                    if (!seen.add(name)) error("duplicate capability \"$name\"")
                    val sensitivity = tbl.getString("sensitivity")
                        ?: error("capability \"$name\": sensitivity missing")
                    if (sensitivity != "general" && sensitivity != "sensitive") {
                        error("capability \"$name\": sensitivity must be \"general\" or \"sensitive\", got \"$sensitivity\"")
                    }
                    add(CapabilityDecl(name, sensitivity))
                }
            }
        }

        return Manifest(
            plugin = PluginSection(id, version, minOrca),
            runtime = RuntimeSection(binary, image, mode, eager),
            surfaces = surfaces,
            capabilities = capabilities,
        )
    }

    fun parseFile(path: Path): Manifest = parseString(Files.readString(path))

    private fun denyUnknown(keys: Set<String>, allowed: Set<String>, where: String) {
        for (k in keys) {
            if (k !in allowed) error("unknown field \"$k\" in $where")
        }
    }

    private fun checkSemver(v: String, field: String) {
        if (v.contains('-') || v.contains('+')) {
            error("$field \"$v\": pre-release/build metadata not supported in v0")
        }
        val parts = v.split('.')
        if (parts.isEmpty()) error("$field \"$v\": empty")
        for (p in parts) {
            if (p.isEmpty()) error("$field \"$v\": empty component")
            if (p.toULongOrNull() == null) error("$field \"$v\": bad numeric component \"$p\"")
        }
    }
}

File: projects/sdk-ts/src/manifest.ts
/**
 * Plugin manifest — `orca-plugin.toml`. Wire-compatible with
 * projects/sdk/src/manifest.rs and projects/sdk-go/manifest.
 */
import { readFile } from 'node:fs/promises';
import { parse as parseToml } from 'smol-toml';

export const FILENAME = 'orca-plugin.toml';

export type Sensitivity = 'general' | 'sensitive';
export type RuntimeMode = 'process';

export interface PluginSection {
  id: string;
  version: string;
  min_orca_version: string;
}

export interface RuntimeSection {
  binary?: string;
  image?: string;
  mode: RuntimeMode;
  eager: boolean;
}

export interface SurfacesSection {
  mcp: boolean;
  cli: boolean;
  ui: boolean;
  docs: boolean;
  jobs: boolean;
  storage: boolean;
  federation: boolean;
}

export interface CapabilityDecl {
  name: string;
  sensitivity: Sensitivity;
}

export interface Manifest {
  plugin: PluginSection;
  runtime: RuntimeSection;
  surfaces: SurfacesSection;
  capabilities: CapabilityDecl[];
}

const ALLOWED_TOP = new Set(['plugin', 'runtime', 'surfaces', 'capabilities']);
const ALLOWED_PLUGIN = new Set(['id', 'version', 'min_orca_version']);
const ALLOWED_RUNTIME = new Set(['binary', 'image', 'mode', 'eager']);
const ALLOWED_SURFACES = new Set([
  'mcp',
  'cli',
  'ui',
  'docs',
  'jobs',
  'storage',
  'federation',
]);
const ALLOWED_CAPABILITY = new Set(['name', 'sensitivity']);

function denyUnknown(obj: Record<string, unknown>, allowed: Set<string>, where: string): void {
  for (const k of Object.keys(obj)) {
    if (!allowed.has(k)) throw new Error(`unknown field "${k}" in ${where}`);
  }
}

function checkSemver(v: string, field: string): void {
  if (v.includes('-') || v.includes('+')) {
    throw new Error(`${field} "${v}": pre-release/build metadata not supported in v0`);
  }
  const parts = v.split('.');
  if (parts.length === 0) throw new Error(`${field} "${v}": empty`);
  for (const p of parts) {
    if (p.length === 0) throw new Error(`${field} "${v}": empty component`);
    if (!/^[0-9]+$/.test(p)) throw new Error(`${field} "${v}": bad numeric component "${p}"`);
  }
}

/** Parse a manifest from a TOML string. */
export function parseString(s: string): Manifest {
  const raw = parseToml(s) as Record<string, unknown>;
  denyUnknown(raw, ALLOWED_TOP, 'top-level');

  const plugin = (raw['plugin'] ?? {}) as Record<string, unknown>;
  denyUnknown(plugin, ALLOWED_PLUGIN, '[plugin]');
  const runtime = (raw['runtime'] ?? {}) as Record<string, unknown>;
  denyUnknown(runtime, ALLOWED_RUNTIME, '[runtime]');
  const surfacesRaw = (raw['surfaces'] ?? {}) as Record<string, unknown>;
  denyUnknown(surfacesRaw, ALLOWED_SURFACES, '[surfaces]');
  const capsRaw = (raw['capabilities'] ?? []) as Array<Record<string, unknown>>;
  for (const c of capsRaw) denyUnknown(c, ALLOWED_CAPABILITY, '[[capabilities]]');

  const id = String(plugin['id'] ?? '').trim();
  if (!id) throw new Error('plugin.id must not be empty');
  for (const ch of id) {
    if (ch === ' ' || ch === '\t' || ch === '\n' || ch === '/' || ch === '\\') {
      throw new Error(
        `plugin.id "${id}" contains invalid characters (whitespace or path separators)`,
      );
    }
  }
  const version = String(plugin['version'] ?? '');
  const minOrca = String(plugin['min_orca_version'] ?? '');
  checkSemver(version, 'plugin.version');
  checkSemver(minOrca, 'plugin.min_orca_version');

  const binary = runtime['binary'] === undefined ? undefined : String(runtime['binary']);
  const image = runtime['image'] === undefined ? undefined : String(runtime['image']);
  if (binary !== undefined && image !== undefined) {
    throw new Error('runtime.binary and runtime.image are mutually exclusive');
  }
  if (binary === undefined && image === undefined) {
    throw new Error('runtime requires either `binary` or `image`');
  }
  const mode = (runtime['mode'] ?? 'process') as string;
  if (mode !== 'process') {
    throw new Error(`runtime.mode "${mode}": only "process" is supported in v0`);
  }
  const eager = Boolean(runtime['eager'] ?? false);

  const surfaces: SurfacesSection = {
    mcp: Boolean(surfacesRaw['mcp'] ?? false),
    cli: Boolean(surfacesRaw['cli'] ?? false),
    ui: Boolean(surfacesRaw['ui'] ?? false),
    docs: Boolean(surfacesRaw['docs'] ?? false),
    jobs: Boolean(surfacesRaw['jobs'] ?? false),
    storage: Boolean(surfacesRaw['storage'] ?? false),
    federation: Boolean(surfacesRaw['federation'] ?? false),
  };

  const capabilities: CapabilityDecl[] = [];
  const seen = new Set<string>();
  for (const c of capsRaw) {
    const name = String(c['name'] ?? '').trim();
    if (!name) throw new Error('capability.name must not be empty');
    if (seen.has(name)) throw new Error(`duplicate capability "${name}"`);
    seen.add(name);
    const sensitivity = String(c['sensitivity'] ?? '');
    if (sensitivity !== 'general' && sensitivity !== 'sensitive') {
      throw new Error(
        `capability "${name}": sensitivity must be "general" or "sensitive", got "${sensitivity}"`,
      );
    }
    capabilities.push({ name, sensitivity });
  }

  const result: Manifest = {
    plugin: { id, version, min_orca_version: minOrca },
    runtime: { mode: 'process', eager, ...(binary !== undefined ? { binary } : {}), ...(image !== undefined ? { image } : {}) },
    surfaces,
    capabilities,
  };
  return result;
}

/** Parse a manifest from a file path. */
export async function parseFile(path: string): Promise<Manifest> {
  const data = await readFile(path, 'utf8');
  return parseString(data);
}
